### PR TITLE
Provide actual literal_eval() function. Adapted from standard library's.

### PR DESCRIPTION
**Pull request by tristanlatr**
Provide actual `literal_eval()` function. Adapted from standard library's.
Use that function in `inference.py` instead of the dummy `_to_literal()` function.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring (tiny bit)  |

### CODE REVIEW

File: astroid/helpers.py
Comment (Pierre-Sassoulas) [inline astroid/helpers.py:342]: We're going to end python 3.6 support soon, but 3.7 will still be used for quite some time.

Comment (tristanlatr) [reply astroid/helpers.py]: Fixed

File: astroid/exceptions.py
Comment (DanielNoord) [inline astroid/exceptions.py:117]: We also re-raise it from `TypeError`, `ValueError` and `LookupError`. So it doesn't seem to only be a `SyntaxError`?

File: astroid/helpers.py
Comment (DanielNoord) [inline astroid/helpers.py:335]:
```suggestion
            raise ValueError(f"Expected only one expression, got {len(_node)}")
```

File: astroid/helpers.py
Comment (DanielNoord) [inline astroid/helpers.py:342]: Would you be willing to check if this is indeed necessary for `astroid` nodes? I would assume we initialise `lineno` on every node, even on `Module` after #1262.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Thanks @tristanlatr. However, I don't really see why it is necessary to maintain our own helper function? If I replace `helpers.literal_eval` in the tests for `ast.literal_eval` it works as well?

Wouldn't it also work if we accepted `str` for `_to_literal` as well and just passed the string instead of `node.as_string()` to `ast.literal_eval`?

File: astroid/helpers.py
Comment (tristanlatr) [reply astroid/helpers.py:380]: `set()` is a special case because we can't create empty set other wise. 
I don;t think we should add this kind of support for `dict()` or `list()` since we can use `{}` and `[]`.

File: astroid/helpers.py
Comment (tristanlatr) [reply astroid/helpers.py:342]: For nodes created from scratch it might be necessary.

File: tests/unittest_helpers.py
Comment (tristanlatr) [reply tests/unittest_helpers.py:284]: Python's

File: astroid/exceptions.py
Comment (tristanlatr) [reply astroid/exceptions.py:117]: I don't know, this is out of the scope if this PR anyway.

File: astroid/helpers.py
Comment (tristanlatr) [reply astroid/helpers.py:335]: I'll add a test

**Comment by DanielNoord** [comment]
> > If I replace helpers.literal_eval in the tests for ast.literal_eval it works as well?
> 
> No, astroid nodes and ast nodes are not 100% compatible.

The tests you added do work with `ast.literal_eval`.

> > Wouldn't it also work if we accepted str for _to_literal as well and just passed the string instead of node.as_string() to ast.literal_eval?
> 
> Yes but the cool thing with `helpers.literal_eval` is that we can evaluate `NodeNG` without converting them to string before (which is hightly innefficient)!
>
> I'ts necessary if we don't want to convert back to string all `NodeNG` we might want to evaluate...

Hmm, that's indeed a benefit. However, I wonder if this warrants having to maintain this ourselves. We only use `_to_literal` in the inference of `nodes.Compare`. I'll leave this decision to others though. Perhaps it would be good if we had some actual data just how inefficient `_to_literal` is right now.

File: astroid/helpers.py
Comment (DanielNoord) [reply astroid/helpers.py:342]: That shouldn't be the case. The `init` of each node should handle setting `lineno`. If they do not so this would be a bug and should be fixed.

File: tests/unittest_helpers.py
Comment (DanielNoord) [reply tests/unittest_helpers.py:284]: Perhaps add a reference to it then. Referring to issues from other projects/repo without any context can be confusing I think.

File: astroid/exceptions.py
Comment (DanielNoord) [reply astroid/exceptions.py:117]: Wouldn't the comment then not also be out of scope?

File: astroid/helpers.py
Comment (DanielNoord) [reply astroid/helpers.py:380]: Ah yeah. Still, if we're going to maintain our own function shouldn't we add some obvious extra functionality while we're at it?

Comment (tristanlatr) [reply astroid/helpers.py:380]: I get your point, we could handle empty calls to `list`, `dict`, `tuple`, `frozenset`, if it's only that, I'm ok implementing it. 

But in general, I think we should stay the closest possible from `ast` module, and let users call this function with a possible inferred node.

**Comment by tristanlatr** [comment]
I agree it could work by doing something like:
```python
def _dummy_literal_eval(node: Union[str, NodeNG]) -> Any:
    return ast.literal_eval(node if isinstance(node, str) else node.as_string())
```

But maybe some speed data comparaison will make you prefer my proposition: 
```python
>>> import timeit
>>> timeit.timeit('literal_eval(node)', setup='import json;from pathlib import Path;from astroid.helpers import literal_eval;from astroid.builder import extract_node;node = extract_node(repr(json.load(Path("./10kSamplejson.json").open("r"))))', number=20)
0.9220764310000007
>>> timeit.timeit('_dummy_literal_eval(node)', setup='import json;from pathlib import Path;from astroid.helpers import _dummy_literal_eval;from astroid.builder import extract_node;node = extract_node(repr(json.load(Path("./10kSamplejson.json").open("r"))))', number=20)
7.558820483999995
```
I'v used this big testing json: the JSON file is here https://docs.google.com/uc?export=download&id=1RAp9w12hZkmLt2vJcAjdT-UtC3KpFyq4

File: astroid/helpers.py
Comment (DanielNoord) [reply astroid/helpers.py:380]: Yeah, I think it makes sense to add those three as well. There is a big chance that we will run out of sync with the "upstream" version of this due to limited maintenance time. Makes sense imo to add some slight improvements while we're looking at it.

**Comment by DanielNoord** [comment]
I'd like someone else's opinion on this. I'm -0.5 to it.

Obviously this would be a performance improvement and it might make sense to add this. However, since we're using this only in one place ourselves and we're basically copying the function from `ast` I keep wondering whether the improvement outweighs potentially missing out on any "upstream" fixes. Especially considering the clear lack of active (and more importantly experienced) maintainers we're currently experiencing.
I'm certainly not going to block this as I do see the benefit, but I'd like someone else's opinion on this matter before merging/finalising this!

(Thanks @tristanlatr for the ping, I should have made this clearer 😄)

**Comment by Pierre-Sassoulas** [comment]
Opened #1361 to discuss this.

File: astroid/helpers.py
Comment (tristanlatr) [inline astroid/helpers.py:385]:
```suggestion
            if node.func.name == "set":
                return set()
            elif node.func.name == "tuple":
                return tuple()
            elif node.func.name == "list":
                return list()
            elif node.func.name == "dict":
                return dict()
            else:
                _raise_malformed_node(node)
```

**Review by cdce8p — CHANGES_REQUESTED** [review]
I'm with @DanielNoord on this on. It isn't clear to me why this would be useful. Even if it provides a small performance benefit, the cost of maintaining it is higher IMO. Especially considering `literal_eval` is only used in the inference of `Compare` nodes. With that kind of limited scope, it's unlikely that we would see any performance improvements.

If you do have a use-case for it, I would be interested to hear it. Maybe as response here? https://github.com/PyCQA/astroid/issues/1361#issuecomment-1025275715

**Comment by cdce8p** [comment]
I would like to close this one. As I explained in https://github.com/PyCQA/astroid/issues/1361#issuecomment-1025815781, I don't think the helper would provide value for astroid at the current time. Furthermore, it would increase the maintenance burden if a new Python version is released and something changes.

IMO every time there is an alternative in the stdlib, it's better to use that than to implement your own version for it. In this case the workaround with `ast.literal_eval(node.as_string())` is good enough.

**Comment by DanielNoord** [comment]
Seeing as this discussion seems to have died down I think I'm going to close this. There seems to be somewhat of a consensus (from the maintainer's side) that this is not something we would want to add right now.

**Pull request closed without merging**
